docs/service-add-ons.md

```diff
@@ -0,0 +1,170 @@
+affecting the service as a whole, e.g.: logging.
+
+The configuration base type won’t share any common fields. Existing ad-hoc 
+components use endpoint but it is possible to have some other add-ons that do 
+not use that configuration at all.
+
+The factory will follow the pattern established for pipeline configuration:
+
+```go
+// Factory is a factory interface for add-ons to the service.
+type Factory interface {
+    // Type gets the type of the add-on component created by this factory.
+    Type() string 
+
+    // CreateDefaultConfig creates the default configuration for the add-on.
+    CreateDefaultConfig() configmodels.AddOn 
+
+    // CustomUnmarshaler returns a custom unmarshaler for the configuration or nil if
+    // there is no need for custom unmarshaling. This is typically used if viper.Unmarshal()
+    // is not sufficient to unmarshal correctly.
+    CustomUnmarshaler(v *viper.Viper, viperKey string, intoCfg interface{}) CustomUnmarshaler 
+
+    // CreateAddOn creates a service add-on based on the given config.
+    CreateAddOn(logger *zap.Logger, cfg configmodels.AddOn) (Component, error)
+}
+```
+
+
+## Add-On Component Interface
+
+The interface defined below is the minimum required to keep same behavior for 
+ad-hoc components currently in use on the service:
+
+```go
+// Component is the interface for objects hosted by the OpenTelemetry Service that
+// doesn't participate directly on data pipelines but provide some functionality
+// to the service, examples: health check endpoint, z-pages, etc.
+type Component interface {
+    // Start the Component object hosted by the given host. At this point in the
+    // process life-cycle the receivers are not started and the host did not
+    // receive any data yet.
+    Start(host Host) error 
+
+    // Shutdown the Component instance. This happens after the pipelines were
+    // shutdown.
+    Shutdown() error
+}
+
+// PipelineWatcher is an extra interface for Components hosted by the OpenTelemetry
+// Service that is to be implemented by Components interested in changes to pipeline
+// states. Typically this will be used by Components that change their behavior if data is
+// being ingested or not, e.g.: a k8s readiness probe.
+type PipelineWatcher interface {
+    // Ready notifies the Component that all pipelines were built and the
+    // receivers were started, i.e.: the service is ready to receive data
+    // (notice that it may already have received data when this method is called).
+    Ready() error 
+
+    // NotReady notifies the Component that all receivers are about to be stopped,
+    // i.e.: pipeline receivers will not accept new data.
+    // This is sent before receivers are stopped, so the Component can take any
+    // appropriate action before that happens.
+    NotReady() error
+}
+
+// Host represents the entity where the add-on is being hosted. It is used to
+// allow communication between the add-on and its host.
+type Host interface {
```

Comment:
I think we already have a `Host` interface in the codebase, right?

Reply:
We do have a `receiver.Host` interface (with a single different method). I'm not sure if we should coalesce them into a single one or not. Separately we can trim the complexity of the host and make the "add-ons" and receivers easier to reuse (host just need to implement what is needed for each different interface).

Another alternative is to make a kind of basic host and move what is common to all host types there.

Reply:
Both `Host`s are currently used for error reporting, but receiver also needs a Context, not sure if Context makes sense for Addons, I didn't think though this in details, just wanted to make sure you intentionally created the new Host interface.